azurerm/internal/tools/website-scaffold/main.go

```diff
@@ -0,0 +1,891 @@
+	brandName string
+
+	// resourceName is the name of the resource e.g. `azurerm_resource_group`
+	resourceName string
+
+	// isDataSource defines if this is a Data Source (if not it's a Resource)
+	isDataSource bool
+
+	// resourceId is an example of the ID used by this Resource
+	resourceId *string
+
+	// websiteCategories is the list of categories available by this service definition
+	websiteCategories []string
+}
+
+func (gen documentationGenerator) generate() string {
+	title := gen.title()
+	argumentsBlock := gen.argumentsBlock()
+	attributesBlock := gen.attributesBlock()
+	description := gen.description()
+	exampleUsageBlock := gen.exampleUsageBlock()
+	frontMatterBlock := gen.frontMatterBlock()
+	importBlock := gen.importBlock()
+	timeoutsBlock := gen.timeoutsBlock()
+
+	template := fmt.Sprintf(`%s
+# %s
+
+%s.
+
+## Example Usage
+
+[][][]hcl
+%s
+[][][]
+
+%s
+
+%s
+
+%s
+
+%s`, frontMatterBlock, title, description, exampleUsageBlock, argumentsBlock, attributesBlock, timeoutsBlock, importBlock)
+	return strings.ReplaceAll(template, "[][][]", "```")
+}
+
+// blocks
+func (gen documentationGenerator) argumentsBlock() string {
+	var documentationForArguments = func(input map[string]*schema.Schema, onlyRequired, onlyOptional bool, blockName string) string {
+		fields := ""
+
+		for _, fieldName := range gen.sortFields(input) {
+			field := input[fieldName]
+
+			// nothing to see here, move along
+			if !field.Optional && !field.Required {
+				continue
+			}
+
+			if onlyRequired && !field.Required {
+				continue
+			}
+
+			if onlyOptional && !field.Optional {
+				continue
+			}
+
+			status := "Optional"
+			if field.Required {
+				status = "Required"
+			}
+
+			value := gen.buildDescriptionForArgument(fieldName, field, blockName)
+			if len(field.ConflictsWith) > 0 {
+				conflictingValues := make([]string, 0)
+				for _, v := range field.ConflictsWith {
+					conflictingValues = append(conflictingValues, fmt.Sprintf("`%s`", v))
+				}
+
+				value += fmt.Sprintf("Conflicts with %s", strings.Join(conflictingValues, ","))
+			}
+			if field.ForceNew {
+				value += fmt.Sprintf(" Changing this forces a new %s to be created.", gen.brandName)
+			}
+			fields += fmt.Sprintf("* `%s` - (%s) %s\n\n", fieldName, status, value)
+		}
+
+		return fields
+	}
+
+	// first output the Required fields
+	fields := documentationForArguments(gen.resource.Schema, true, false, "")
+	// then prepare the Optional fields
+	optionalFields := documentationForArguments(gen.resource.Schema, false, true, "")
+
+	// assuming we have both optional & required fields - let's add a separarer
+	if len(fields) > 0 && len(optionalFields) > 0 {
+		fields += "---\n\n"
+	}
+	fields += optionalFields
+
+	// first list all of the top-level fields / blocks alphabetically
+
+	// then we need to collect a list of all block names, everywhere
+	blockNames, blocks := gen.uniqueBlockNamesForArgument(gen.resource.Schema)
```

Comment:
Given the case that there are two arguments sharing the same block type, then it will be repeated in the document, which seems redundent.

Barely comparing schema definition to predict whether two arguments share a same block type is not so reliable, however,  in this case maybe we can add a **TODO** comment to remind user to choose.

Reply:
we're scaffolding rather than generating - thus this is an acceptable limitation